src/contracts/middleware/IndexRegistry.sol

```diff
@@ -0,0 +1,263 @@
+        if (index != _operatorIdToIndexHistory[operatorId][quorumNumber].length - 1) {
+            OperatorIndexUpdate memory nextOperatorIndex = _operatorIdToIndexHistory[operatorId][quorumNumber][index + 1];
+            require(blockNumber < nextOperatorIndex.fromBlockNumber, "IndexRegistry.getOperatorIndexForQuorumAtBlockNumberByIndex: provided index is too far in the future for provided block number");
+        }
+        return operatorIndexToCheck.index;
+    }
+
+    /**
+     * @notice Looks up the number of total operators for `quorumNumber` at the specified `blockNumber`.
+     * @param quorumNumber is the quorum number for which the total number of operators is desired
+     * @param blockNumber is the block number at which the total number of operators is desired
+     * @param index is the index of the entry in the dynamic array `_totalOperatorsHistory[quorumNumber]` to read data from
+     */
+    function getTotalOperatorsForQuorumAtBlockNumberByIndex(uint8 quorumNumber, uint32 blockNumber, uint32 index) external view returns (uint32){
+        OperatorIndexUpdate memory operatorIndexToCheck = _totalOperatorsHistory[quorumNumber][index];
+
+        // blocknumber must be at or after the "index'th" entry's fromBlockNumber
+        require(blockNumber >= operatorIndexToCheck.fromBlockNumber, "IndexRegistry.getTotalOperatorsForQuorumAtBlockNumberByIndex: provided index is too far in the past for provided block number");
+        
+        // if there is an index update after the "index'th" update, the blocknumber must be before the next entry's fromBlockNumber
+        if (index != _totalOperatorsHistory[quorumNumber].length - 1){
+            OperatorIndexUpdate memory nextOperatorIndex = _totalOperatorsHistory[quorumNumber][index + 1];
+            require(blockNumber < nextOperatorIndex.fromBlockNumber, "IndexRegistry.getTotalOperatorsForQuorumAtBlockNumberByIndex: provided index is too far in the future for provided block number");
+        }
+        return operatorIndexToCheck.index;
+    }
+
+    /// @notice Returns an ordered list of operators of the services for the given `quorumNumber` at the given `blockNumber`
+    function getOperatorListForQuorumAtBlockNumber(uint8 quorumNumber, uint32 blockNumber) external view returns (bytes32[] memory){
+        bytes32[] memory quorumOperatorList = new bytes32[](_getTotalOperatorsForQuorumAtBlockNumber(quorumNumber, blockNumber));
+        for (uint i = 0; i < globalOperatorList.length; i++) {
+            bytes32 operatorId = globalOperatorList[i];
+            uint32 index = _getIndexOfOperatorForQuorumAtBlockNumber(operatorId, quorumNumber, blockNumber);
+            // if the operator was not in the quorum at the given block number, skip it
+            if (index == OPERATOR_DEREGISTERED_INDEX)
+                continue;
+            quorumOperatorList[index] = operatorId;
+        }
+        return quorumOperatorList;
+    }
```

Comment:
any concerns about the unbounded nature of this loop?
I guess the function can always be simulated with eth_call and an excessively high gas budget?